minimal_scheduling/minimal_sched_fifo_callback_group.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pthread.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+/* For this example, we will be using multiple callback groups to separate callbacks, so it is
+ * possible to spin them in threads with different scheduling priorities. */
+
+using namespace std::chrono_literals;
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count1_(0), count2_(0)
+  {
+    publisher1_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback1 =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count1_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing     : '%s'", message.data.c_str());
+        this->publisher1_->publish(message);
+      };
+    timer1_ = this->create_wall_timer(500ms, timer_callback1);
+
+    realtime_callback_group_ = this->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+    publisher2_ = this->create_publisher<std_msgs::msg::String>("topic_rt", 10);
+    auto timer_callback2 =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count2_++);
```

Comment:
`timer_callback2` is RT, right? Doesn't `std::to_string` dynamically allocate?

Reply:
That's right. In general, in these examples, I don't pretend the code we show is real-time compliant. I just want to show one specific concept. In this case, std::string is just one of the problems, `std_msgs::msg::String()` will allocate too.

Maybe I can add a comment explaining this is not real-time or something like this.

Reply:
To give you some context, I'm using the minimal publishers and subscribers from [ros2/examples](https://github.com/ros2/examples/tree/master/rclcpp/topics) with no or minimal modifications. In order to avoid confussion for users I see the following options:

1. Leave it as it is and add a note explaining this is not real-time compliant code
2. Replace `std_msgs::msg::String` with some POD message type `std_msgs::msg::Int32` and remove the logging macros. Now we could claim a callback is meant to be real-time compliant.
3. Rename some variables so we describe better what we are doing here. It's not necessarily "real-time" in this example, is has a higher priority (i.e. realtime_callback_group_ => high_prio_callback_group_)

FYI @LanderU

Reply:
That makes sense. I think just adding a note in the readme saying this example is not RT and only shows the scheduling config is good enough.